xbmc/addons/AddonManager.cpp

```diff
@@ -230,19 +230,26 @@ VECADDONS CAddonMgr::GetAvailableUpdates() const
   VECADDONS all_addons;
   m_database.GetRepositoryContent(all_addons);
   std::map<std::string, AddonPtr> latest_versions;
+  std::map<std::string, AddonPtr> latest_trusted_versions;
   for (const auto& addon : all_addons)
   {
-    const auto& latest_known = latest_versions.find(addon->ID());
-    if (latest_known == latest_versions.end() || addon->Version() > latest_known->second->Version())
-      latest_versions[addon->ID()] = std::move(addon);
+    if (IsFromTrustedRepo(addon))
+    {
+      MoveAddonToMap(addon, latest_trusted_versions);
+    }
+    else
+    {
+      MoveAddonToMap(addon, latest_versions);
+    }
   }
 
   GetAddonsForUpdate(installed);
   for (const auto& addon : installed)
   {
-    const auto& remote = latest_versions.find(addon->ID());
-    if (remote != latest_versions.end() && remote->second->Version() > addon->Version())
-      updates.emplace_back(std::move(remote->second));
+    if (!CheckVersionAndQueueUpdate(addon, latest_trusted_versions, updates))
+    {
+      CheckVersionAndQueueUpdate(addon, latest_versions, updates);
+    }
```

Comment:
i think that is already the case because we're calling `!CheckVersionAndQueueUpdate` (Line 249) on the latest_trusted_versions.  maybe the function name is misleading?
if i'm completely wrong now `if (!IsFromTrustedRepo(addon))` would do the trick, but i think that's redundant.

Reply:
If the version in `latest_trusted_versions` is not greater, then it will return false I think and then try the private repos.

You could change the approach slightly and have `latest_versions` be a superset, so it contains all latest versions: private + trusted.

Then you code could be:

```cpp
if (!IsFromTrustedRepo(addon))
  CheckVersionAndQueueUpdate(addon, latest_trusted_versions, updates);
else
  CheckVersionAndQueueUpdate(addon, latest_versions, updates);
```

Maybe the bit of redundancy makes the code cleaner, what do you think?

Reply:
Or if you want to keep the maps mutually exclusive:

```cpp
if (IsFromTrustedRepo(addon))
{
  CheckVersionAndQueueUpdate(addon, latest_trusted_versions, updates);
}
else
{
  if (CheckVersionAndQueueUpdate(addon, latest_trusted_versions, updates))
    CheckVersionAndQueueUpdate(addon, latest_versions, updates);
}
```

Reply:
Although that does mean probably just extending your original code is best!

```cpp
    if (!CheckVersionAndQueueUpdate(addon, latest_trusted_versions, updates) && !IsFromTrustedRepo(addon))
    {
      CheckVersionAndQueueUpdate(addon, latest_versions, updates);
    }
```

Reply:
hold on ;) that's a great idea...
atm. `CheckVersionAndQueueUpdate` returns false if the addon is not found in the queried map.
would be better if it just returned if an update is queued or not. then we could use your cleanly readable call:
```
if (IsFromTrustedRepo(addon))
  CheckVersionAndQueueUpdate(addon, latest_trusted_versions, updates);
else
  CheckVersionAndQueueUpdate(addon, latest_versions, updates);
``` 
even though the return-value is not useful here (but maybe later)

i'm just not sure if we then need a superset-map. imho we have to kepp them split:
`latest_trusted_versions` contains repository.xbmc.org and others...
`latest_versions` contains 3rd party stuff
i would do it that way

Reply:
But I think you still need to check trusted first followed by private then for private repo addons.

Reply:
that's right and needs a little adjustment

Reply:
hm, sorry, i don't get it right:

let's say we have installed `repository.xbmc.org` and matt's `repository.sticky`.
both repos contain `service.xbmc.versioncheck`, so each of our two maps is holding an entry for this addon.

`IsFromTrustedRepo(addon)` - returns true, so we only check the trusted map for updates and do nothing more.

- now the private addon: `script.anything` is **not** present in `latest_trusted_versions`.. (it can't be, it doesn't exist in  `repository.xbmc.org`.   -- so i can't understand why we should need to check the trusted first. what am i missing?

Reply:
I get you. Yes, you are right. The condition is the check. Nice